### PR TITLE
Refactorings

### PR DESCRIPTION
**Pull request by karyon**
see commit messages. also includes #643 with changed commit message because why not.

closes #519

### CODE REVIEW

**Comment by karyon** [comment]
@diox, the "remove silent exception catching" commit revealed randomly failing test. could you have a look at that? i'm not sure where to start, i can't repro this locally.

here's another test run of the same commit: https://travis-ci.org/django-compressor/django-compressor/builds/80111522

**Comment by diox** [comment]
I've restarted the builds that errored but it's failing again. I have no idea why this test fails, I've never seen it fail before...

**Comment by diox** [comment]
I'd prefer it this PR was splitted into multiple smaller ones, with tests covering the changes. Otherwise I have no idea what I'm merging or why.

**Comment by diox** [comment]
Note: if you rebase, you should benefit from #656 which makes the tests run on travis sudo-less architecture, which means faster builds, which means faster failures to help debugging this :)

**Comment by karyon** [comment]
the failures appear to be random. i've restarted the tests about 4 times, and it's always different runs that are failing, but always the same test and error.

you haven't seen this failure before because there was a "except Exception" which hid the problem. i removed the except in a commit.

opening a PR for every single commit seemed a bit overkill to me. but if you really want, i can do that later.

**Comment by diox** [comment]
I see. Can you just isolate "remove silent exception catching" in a separate PR ? It'll make it easier to merge the rest first, and then we can concentrate on fixing this random failure.

**Comment by karyon** [comment]
alright, i completely separated the commits now. please let me know whether you want anything changed.

File: compressor/base.py [hunk 1]
```diff
@@ -36,17 +36,18 @@ class Compressor(object):
     Base compressor object to be subclassed for content type
     depending implementations details.
     """
-    type = None
 
-    def __init__(self, content=None, output_prefix=None, context=None, *args, **kwargs):
+    def __init__(self, content=None, output_prefix=None, context=None, filters=None):
```
Comment (diox) [inline compressor/base.py]: I prefer leaving `*args` and (more importantly) `**kwargs`, removing them is a backwards-incompatible change for people inheriting from it.

**Comment by karyon** [comment]
done.

**Comment by diox** [comment]
I'm not a fan of https://github.com/karyon/django-compressor/commit/24551f5dbd28084077eb17b9291b6140fb21de03. It removes methods that are useful for users inheriting from compressor. What was the motivation behind this change ?

**Comment by karyon** [comment]
i removed render_cached because there was only one caller and it removing it reduced code complexity noticably. render_output just seemed superfluous so i inlined it, debug_mode was called nowhere so i simply removed it.

i can absolutely restore them if you want, although for render_cached i'd need to check again whether i can keep some of the simplifications, and for debug_mode i'd like to hear when it is intended to be called :)

**Comment by diox** [comment]
I kinda like having `render_cached` separated so that it's easy to implement custom behaviour by inheriting from the class (the `is_offline_compression_enabled` exists for a similar purpose, I've used in the the past to make custom compress template tag nodes that are not offline cached, despite behaving normally otherwise).

The rest, and in particular the extra `debug_mode` method, can go.

**Comment by karyon** [comment]
i split up the respective commit and added a deprecation warning fix. please review.

basically, the refactoring moves the decision whether to render (cached or offline) from the methods into the main method, which makes things a bit easier to read (for me at least)

**Comment by karyon** [comment]
i think https://github.com/karyon/django-compressor/commit/68981ebfbf8af2a47a24a85fcee0d660a556d529#diff-ac8de32c7b59ab69b2ac7137529224f0R85 also kinda includes https://github.com/django-compressor/django-compressor/pull/519.

diox, could you confirm that the offline_manifest[key] cannot return None when key in offline_manifest?

also, since these "refactorings" apparently do fix something, do you think a test is necessary?

**Comment by diox** [comment]
`offline_manifest[key]` should not return `None`. It's a simple dict (stored and loaded using a JSON file).

Since you're slightly changing behaviour (for the better), consider adding a test, yes.

**Comment by karyon** [comment]
test added.

**Comment by diox** [comment]
Excellent, thanks.

**Pull request merged**